IES prediction on Loxodes MIC and MAC genomes
=============================================

BleTIES MILRAA using CCS reads cross-mapped onto Loxodes magnus Falcon
reference assemblies.

Snakemake workflow files are in subfolder `./workflow`; Snakemake options are
in the script `run_snakemake.sh`

Input files for this workflow are genome assemblies and read mappings for L.
magnus, listed in `workflow/config.yml`. Workflow to produce these files:
[loxodes-assembly-workflow](https://github.com/Swart-lab/loxodes-assembly-workflow).

Input data are available from:
"Loxodes magnus genome assemblies and annotations",
https://doi.org/10.17617/3.9QTROS, Edmond.

Results of variant calling are available from:
"Loxodes magnus indel polymorphisms and variant calling",
https://doi.org/10.17617/3.NEV8C1, Edmond.


Related datasets
----------------

 * Genomic sequencing for sorted MIC vs. MAC of Loxodes magnus: European
   Nucleotide Archive [PRJEB55123](https://www.ebi.ac.uk/ena/browser/view/PRJEB55123)


Contact
-------

[Brandon Seah](mailto:kb.seah@tuebingen.mpg.de)

[Estienne Swart](mailto:estienne.swart@tuebingen.mpg.de)
